proposals/0033-timely-vote-credits.md

```diff
@@ -0,0 +1,389 @@
+credits are already guaranteed by existing VoteStateUpdate correctness
+checking code:
+
+The existing VoteStateUpdate processing code already ensures that (1) only
+roots slots that could actually have been rooted in the transition from the
+old VoteState to the new VoteState, so there is no danger of over-counting
+credits (i.e. imagine that a 'cheating' validator "pretended" that slots
+were rooted by dropping them off of the back of the new VoteState before
+they have actually achieved 32 confirmations; the existing logic prevents
+this).
+
+The existing VoteStateUpdate processing code already ensures that (2) new
+slots included in the new VoteState are only slots after slots that have
+already been voted on in the old VoteState (i.e. can't inject new slots in
+the middle of slots already voted on).
+
+## Impact
+
+1. Any code which deserialized vote account state will require a recompile
+   with the newest Solana SDK in order to be able ot deserialize the new
+   vote account state.
+
+2. There will be a strong anti-incentive for vote lagging.  Validators who
+   currently delay their votes in order to reduce their chances of being
+   locked out from voting, will no longer do so.
+
+3. The Solana CLI will have been updated to show vote latencies for each
+   vote; any context in which the Solana CLI output is observed by users
+   or other tooling will need to adapt to the additonal data being
+   displayed.
+
+## Security Considerations
+
+The are no security considerations; however there are changes to incentive
+structure for voting which may result in differing vote behavior becoming
+more optimal with regards to earning credits.
+
+To maximize vote credits, a validator must vote on all slots that end up
+being rooted by the cluster, and no slots that are not rooted.  This means
+always "picking the right fork" whenever there are forks.  It is the drive
+to always vote on the correct fork that causes validators to choose a vote
+lagging strategy in the first place; they can earn more credits if they
+reduce their chances of voting on "the wrong fork" by waiting until
+sufficient votes from other validators have landed on a given slot before
+voting on it.
+
+Therefore the optimal vote strategy currently is to wait until a slot is
+rooted (i.e. has received 31 confirmations) before voting on it.  In this
+way, perfect vote credits would be earned.  However, this optimal voting
+strategy, if employed cluster-wide, would halt the cluster as no validators
+would vote, all of them waiting for others to vote first.
+
+It's possible to get very close to this optimal strategy by waiting for some
+number of confirmations less than 31, with the number of incidents of
+picking the wrong fork increasing as the waiting time is decreased.  The
+current "vote laggers" have done this, with varying degrees of skill.
+
+After the timely vote credits change, the dynamics will change; it will no
```

Comment:
These three paragraphs would be better in the impact section, the impact on the new optimal voting strategy.

Reply:
Again the "Security Considerations" section was meant to specifically address issues that have already been brought up several times in validator meetings to discuss this feature; having identified that as a point that people generally have concern about, I thought it best to address that topic very clearly here.